runtime/android/core/src/org/xwalk/core/XWalkActivity.java

```diff
@@ -91,99 +71,53 @@
-            mXWalkActivity = activity;
-        }
-
-        @Override
-        public void onDecompressionStarted() {
-            mXWalkActivity.showDialog(mXWalkActivity.getDecompressionProgressDialog());
-        }
-
-        @Override
-        public void onDecompressionCancelled() {
-            mXWalkActivity.dismissDialog();
-            mXWalkActivity.finish();
-        }
-
-        @Override
-        public void onDecompressionCompleted() {
-            mXWalkActivity.dismissDialog();
-            mXWalkActivity.initXWalkLibrary();
-        }
-
-        @Override
-        public void onDownloadStarted() {
-            mXWalkActivity.showDialog(mXWalkActivity.getDownloadProgressDialog());
-        }
-
-        @Override
-        public void onDownloadUpdated(int percentage) {
-            ProgressDialog dialog = (ProgressDialog) mXWalkActivity.mActiveDialog;
-            dialog.setIndeterminate(false);
-            dialog.setMax(100);
-            dialog.setProgress(percentage);
-        }
-
-        @Override
-        public void onDownloadCancelled() {
-            mXWalkActivity.dismissDialog();
-            mXWalkActivity.finish();
-        }
-
-        @Override
-        public void onDownloadCompleted(Uri uri) {
-            mXWalkActivity.dismissDialog();
-
-            Log.d(TAG, "Install the Crosswalk library, " + uri.toString());
-            Intent install = new Intent(Intent.ACTION_VIEW);
-            install.setDataAndType(uri, "application/vnd.android.package-archive");
-            mXWalkActivity.startActivity(install);
-        }
+    private boolean mBackgroundDecorated;
 
-        @Override
-        public void onDownloadFailed(int status, int error) {
-            mXWalkActivity.dismissDialog();
-            mXWalkActivity.showDialog(mXWalkActivity.getDownloadFailedDialog(status, error));
-        }
-
-        @Override
-        public void onInitializationStarted() {
-        }
+    /**
+     * This method will be invoked when the Crosswalk environment is ready
+     */
+    protected abstract void onXWalkReady();
 
-        @Override
-        public void onInitializationCompleted() {
-            mXWalkActivity.mIsXWalkReady = true;
-            mXWalkActivity.onXWalkReady();
-        }
+    /**
+     * Returns true if the Crosswalk environment is ready, false otherwise
+     */
+    protected boolean isXWalkReady() {
+        return mIsXWalkReady;
     }
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
 
-        mLibraryListener = new XWalkLibraryListener(this);
-        XWalkLibraryLoader.prepareToInit();
-        XWalkLibraryLoader.startDecompression(mLibraryListener, this);
+        mLibraryDelegate = new XWalkLibraryDelegate();
     }
 
     @Override
     protected void onResume() {
         super.onResume();
-        if (!mIsXWalkReady && !(mActiveDialog instanceof ProgressDialog)) initXWalkLibrary();
```

Comment:
Could move all the code in onResume into onXWalkLibraryPrepared, then mIsXWalkReady is useless.